net: dns: Add DNS cache for improved performance

kica-z opened the pull request:
In order to reduce dns resolve requests when using the dns resolver i added a caching mechanism. I would appreciate some feedback as to if a cache is even something we should add as well as opinions regarding implementation. I have described the current implementation as well as known problems here.

**Implementation:**
The current implementation is as follows:
- One cache for all DNS resolver contexts
- The cache returns as many results as possible for each query
- Multiple queries per entry are supported
- Entries are retained based on their RR TTL and removed as soon as it runs out
- If the cache is full the oldest entry is replaced with the new addition

**Known limitations:**
- When the resolver supports multiple addrinfo per query (CONFIG_DNS_RESOLVER_AI_MAX_ENTRIES) and those entries have different TTLs the cache will at some point no longer return all entries due to TTL expiry. This could be the expected behaviour but also might not be what the user wants.
- If multiple DNS resolver contexts run the same query in parallel duplicate entries could be introduced into the cache since the same address will be added twice.

jukkar commented on tests/net/lib/dns_cache/src/main.c:
```diff
@@ -0,0 +1,136 @@
+
+ZTEST(net_dns_cache_test, test_simple_cache_entry)
+{
+	struct dns_addrinfo info_write = {.ai_family = 42};
```
Is the 42 value here just some arbitrary value, why not use proper address family values?

kica-z replied:
It is not arbitrary (https://en.wikipedia.org/wiki/Phrases_from_The_Hitchhiker%27s_Guide_to_the_Galaxy#Answer_to_the_Ultimate_Question_of_Life,_the_Universe,_and_Everything_(42)) but definitely only a sentinel value 😅.

I have replaced it with a proper address family.

kica-z commented on include/zephyr/net/dns_resolve.h:
```diff
@@ -82,6 +82,13 @@ enum dns_query_type {
+ * label octets and label length octets) is restricted to 255 octets or
+ * less.
+ */
+#define DNS_MAX_NAME_LEN	255
```
A Kconfig option would surely be a good solution however this was already defined within `dns_resolve.c` thats why I reused it. Do you think this can be replaced in both files with this new kconfig option?

jukkar replied:
Hmm, that is a good question. I am a bit worried that the value might be too large for normal use and users might want to tweak it to save memory. So if we have a Kconfig option it should be used in all over the place.
@rlubos what is you opinion about this?

kica-z replied:
@jukkar I added the Kconfig option for now it would not be a big change to revert this.

kica-z commented on include/zephyr/net/dns_resolve.h:
```diff
@@ -82,6 +82,13 @@ enum dns_query_type {
 
```
Done 👍

jukkar replied:
This one PR is good, no need to have multiple PRs.

kica-z commented on subsys/net/lib/dns/Kconfig:
```diff
@@ -128,6 +128,27 @@ module-str = Log level for DNS resolver
 
+menuconfig DNS_RESOLVER_CACHE
+	bool "DNS resolver cache"
+	default n
```
Done 👍

kica-z commented on subsys/net/lib/dns/dns_cache.c:
```diff
@@ -0,0 +1,162 @@
+#include <zephyr/net/dns_resolve.h>
+#include "dns_cache.h"
+
+LOG_MODULE_REGISTER(net_dns_cache, 4);
```
Thanks for pointing this out. I must have forgotten to adjust this after debugging.

kica-z commented on subsys/net/lib/dns/dns_cache.c:
```diff
@@ -0,0 +1,162 @@
+	}
+
+	if (strlen(query) >= DNS_MAX_NAME_LEN) {
+		LOG_WRN("Query string to big to be processed %u >= DNS_MAX_NAME_LEN",
```
I replaced the normal log macros. A question out of interest, what is the difference? Both only show when NET_LOG is activated since `CONFIG_DNS_RESOLVER_LOG_LEVEL` only gets set when this is the case. So I am wondering if there is another difference. Sorry if this might be a stupid question.

jukkar replied:
The NET_LOG macros are mostly same as normal logging macros, but for example NET_DBG prints the thread name or id which normal logging macro does not by default. The idea with separate networking logging macros is that we can adjust the printouts more easily for network specific stuff if needed.

kica-z commented on subsys/net/lib/dns/resolve.c:
```diff
@@ -1184,6 +1186,23 @@ int dns_resolve_name(struct dns_resolve_context *ctx,
+#ifdef CONFIG_DNS_RESOLVER_CACHE
+	struct dns_addrinfo cached_info[CONFIG_DNS_RESOLVER_AI_MAX_ENTRIES] = {0};
+
+	ret = dns_cache_find(&dns_cache, query, cached_info, sizeof(cached_info));
+	if (ret > 0) {
+		/* The query was cached, no
+		 * need to continue further.
+		 */
+		for (size_t cache_index = 0; cache_index < ret; cache_index++) {
+			cb(DNS_EAI_INPROGRESS, &cached_info[cache_index], user_data);
+		}
+		cb(DNS_EAI_ALLDONE, NULL, user_data);
+
+		return 0;
+	}
```
I totally agree with you that this is a very simple implementation. Also as mentioned in the PR notes there are already some known limitations. That is the reason why I would appreciate some feedback on this in order to evaluate what the needs are and how much additional complexity we need.

jukkar replied:
Ok, we can go on with this minimalistic approach first and see if anything else needs to be added. One important thing that I think we need to add with this first version is negative caching, meaning adding information that some dns name was not found.

jukkar commented on subsys/net/lib/dns/dns_cache.c:
```diff
@@ -0,0 +1,162 @@
+	}
+
+	if (strlen(query) >= DNS_MAX_NAME_LEN) {
+		NET_WARN("Query string to big to be processed %u >= DNS_MAX_NAME_LEN",
```
If we set the max length behind the kconfig option, then we need to print the value here instead of the name.

kica-z replied:
Adjusted

jukkar commented on subsys/net/lib/dns/dns_cache.c:
```diff
@@ -0,0 +1,162 @@
+{
+	NET_DBG("REMOVE all entries with query \"%s\"", query);
+	if (strlen(query) >= DNS_MAX_NAME_LEN) {
+		NET_WARN("Query string to big to be processed %u >= DNS_MAX_NAME_LEN",
```
ditto

kica-z replied:
Adjusted

jukkar commented on subsys/net/lib/dns/dns_cache.c:
```diff
@@ -0,0 +1,162 @@
+		return -EINVAL;
+	}
+	if (strlen(query) >= DNS_MAX_NAME_LEN) {
+		NET_WARN("Query string to big to be processed %u >= DNS_MAX_NAME_LEN",
```
and here too

kica-z replied:
Adjusted

jukkar commented on subsys/net/lib/dns/dns_cache.h:
```diff
@@ -0,0 +1,100 @@
+ * @param name Name of the cache.
+ */
+#define DNS_CACHE_DEFINE(name, cache_size)                                                         \
+	K_MUTEX_DEFINE(name##_mutex);                                                              \
```
This could be static too

kica-z replied:
Done

jukkar commented on subsys/net/lib/dns/dns_cache.h:
```diff
@@ -0,0 +1,100 @@
+ */
+#define DNS_CACHE_DEFINE(name, cache_size)                                                         \
+	K_MUTEX_DEFINE(name##_mutex);                                                              \
+	static struct dns_cache_entry name##_entries[cache_size] = {0};                            \
```
No need to initialize to 0 because this is a static variable.

kica-z replied:
Done

kica-z commented on subsys/net/lib/dns/dns_cache.c:
```diff
@@ -0,0 +1,162 @@
+
+	LOG_DBG("ADD \"%s\" with TTL %" PRIu32, query, ttl);
+
+	dns_cache_clean(cache);
```
True but then we would need a loop more as soon as the cache is full every time we add something (without calling find). Do you think this would still be more efficient? Or is there something I am missing here?
```c
        // Additional loop to check if space is available
	for (size_t i = 0; i < cache->size; i++) {
		if (!cache->entries[i].in_use) {
			index_to_replace = i;
			found_empty = true;
		}
	}

	if(!found_empty) {
		dns_cache_clean(cache);
		for (size_t i = 0; i < cache->size; i++) {
			if (!cache->entries[i].in_use) {
				index_to_replace = i;
				found_empty = true;
				break;
			} else if (cache->entries[i].creation_uptime < oldest_entry_uptime) {
				index_to_replace = i;
				oldest_entry_uptime = cache->entries[i].creation_uptime;
			}
		}
	}
```

kica-z commented on subsys/net/lib/dns/resolve.c:
```diff
@@ -635,6 +633,10 @@ int dns_validate_msg(struct dns_resolve_context *ctx,
 			invoke_query_callback(DNS_EAI_INPROGRESS, &info,
 					      &ctx->queries[*query_idx]);
+#ifdef CONFIG_DNS_RESOLVER_CACHE
+			dns_cache_add(&dns_cache,
```
I agree with you that negativ caching would be nice. The question is how to go about this. Normally the TTL for negative caching entries would be set to the value of the `MINIMUM` field of the `SOA`. Currently as far as i know we do not fetch the `SOA` so do you think we should just set a sensible value for the TTL or is it worth it to fetch the `SOA` in order to add a negative cache entry?

rlubos commented on subsys/net/lib/dns/resolve.c:
```diff
@@ -464,14 +440,12 @@ static inline int get_slot_by_id(struct dns_resolve_context *ctx,
-		     int *query_idx,
-		     struct net_buf *dns_cname,
-		     uint16_t *query_hash)
+	int
+	dns_validate_msg(struct dns_resolve_context *ctx, struct dns_msg_t *dns_msg,
```
This indentation looks weird, why is the type indented? Anyway, a general note, it's a bit hard to review due to all those formatting changes, hard to spot the actual meaningful ones. Can we drop it from the PR?

kica-z replied:
Fixed. Removed formatting changes.

rlubos commented on tests/net/lib/dns_cache/on_target.conf:
```diff
@@ -0,0 +1,15 @@
+CONFIG_MAIN_STACK_SIZE=1344
+CONFIG_DNS_RESOLVER=y
+CONFIG_DNS_RESOLVER_CACHE=y
+CONFIG_LOG=y
```
Do we need logs enabled in tests by default? It increases the image size and is not really needed when executing tests. I recall cases from the past, that tests would not fit on some platforms due to logging enabled.

kica-z replied:
Removed log kconfigs

rlubos commented on tests/net/lib/dns_cache/src/main.c:
```diff
@@ -0,0 +1,137 @@
+#include <zephyr/ztest.h>
+#include <zephyr/logging/log.h>
+#include "dns_cache.h"
+LOG_MODULE_REGISTER(net_dns_cache_test, CONFIG_DNS_RESOLVER_LOG_LEVEL);
```
It doesn't seem that logging is used anywhere in the test, please remove it.

rlubos commented on tests/net/lib/dns_cache/prj.conf:
```diff
@@ -0,0 +1,15 @@
+CONFIG_NET_LOG=y
+CONFIG_DNS_RESOLVER_LOG_LEVEL_DBG=y
+
+CONFIG_UBSAN=y
+CONFIG_ASAN=y
```
IMO default `prj.conf` shoud be generic, POSIX-specifc configs should either go in a custom config file, or even better, as  `extra_configs` for posix target in `testcase.yml`.

rlubos commented on subsys/net/lib/dns/dns_cache.c:
```diff
@@ -0,0 +1,174 @@
+	}
+
+	if (found == 0) {
+		NET_DBG("COULD NOT FIND \"%s\"", query);
```
Please don't use all caps in logs.

kica-z replied:
Adjusted

rlubos commented on subsys/net/lib/dns/dns_cache.c:
```diff
@@ -0,0 +1,174 @@
+	}
+
+	if (!found_empty) {
+		dns_cache_clean(cache);
```
Maybe just clean the cache before the first loop, so that we don't need to check for `in_use` twice?

kica-z replied:
Thats what I had before. Reverted to stage where this was the case.

rlubos commented on subsys/net/lib/dns/dns_cache.c:
```diff
@@ -0,0 +1,174 @@
+		if (!cache->entries[i].in_use) {
+			continue;
+		}
+		int64_t reftime = cache->entries[i].creation_uptime;
```
About the time calculations - I wonder that whether it'd make sense, instead of storing creation time + ttl, just have a `k_timepoint_t` for the expiry time, then you could simply use `sys_timepoint_expired()` to check if the entry is still valid, instead of doing the calculations manually.

kica-z replied:
This did however change the functionality a bit since now the entry closest to expiry is the one that will be removed instead of the oldest when additional space is needed.

jukkar commented on include/zephyr/net/dns_resolve.h:
```diff
@@ -89,49 +89,49 @@ enum dns_query_type {
 	struct sockaddr ai_addr;
-	socklen_t       ai_addrlen;
-	uint8_t            ai_family;
-	char            ai_canonname[DNS_MAX_NAME_SIZE + 1];
+	socklen_t ai_addrlen;
+	uint8_t ai_family;
+	char ai_canonname[DNS_MAX_NAME_SIZE + 1];
```
This seems to be unrelated change, please remove this, or add this to new commit if you think this should be there.

jukkar commented on include/zephyr/net/dns_resolve.h:
```diff
@@ -89,49 +89,49 @@ enum dns_query_type {
  */
 enum dns_resolve_status {
 	/** Invalid value for `ai_flags' field */
-	DNS_EAI_BADFLAGS    = -1,
```
Same for these cosmetic changes.

jukkar commented on include/zephyr/net/dns_resolve.h:
```diff
@@ -175,10 +174,10 @@ struct dns_resolve_context {
-		uint8_t is_mdns : 1;
+		uint8_t is_mdns: 1;
 
 		/** Is this server LLMNR one */
-		uint8_t is_llmnr : 1;
+		uint8_t is_llmnr: 1;
```
and here too. In fact it seems that all the changes to `dns_resolve.h` are cosmetic and should be removed from this PR.

jukkar commented on tests/net/lib/dns_cache/CMakeLists.txt:
```diff
@@ -0,0 +1,10 @@
+# SPDX-License-Identifier: Apache-2.0
```
Please change the subject of test commit to `tests: net: dns: ....`

jukkar reviewed (commented):
Looks pretty good, please check the test failure in CI.

kica-z commented:
@jukkar I addressed your newest comments (still investigating regarding CI).
Regarding negative caching should we implement this now or later? I mentioned in my https://github.com/zephyrproject-rtos/zephyr/pull/68825#discussion_r1485928466 that the `SOA` ist still unclear.

kica-z committed:
net: dns: Add DNS cache for improved performance

In order to reduce dns resolve requests when using the dns
resolver an optional cache was introduced. This cache
retains query requests for the TTL duration and therefore
prevents premature refetching of DNS RRs.

Signed-off-by: Carlo Kirchmeier <carlo.kirchmeier@zuehlke.com>

kica-z committed:
tests: net: dns: Add DNS cache tests

Added tests in order to verify the functionality of the newly
added DNS cache.

Signed-off-by: Carlo Kirchmeier <carlo.kirchmeier@zuehlke.com>

jukkar commented:
> Regarding negative caching should we implement this now or later?

We can do it later too.

kica-z commented:
@jukkar Build error has been fixed. Let me know if something still needs to be done.

kica-z commented on tests/net/lib/dns_cache/src/main.c:
```diff
@@ -0,0 +1,135 @@
+		"Cache entry adding should work.");
+	zassert_equal(3, dns_cache_find(&test_dns_cache, query, info_read, 3));
+	zassert_equal(AF_INET, info_read[0].ai_family);
+	k_sleep(K_MSEC(TEST_DNS_CACHE_DEFAULT_TTL * 1000 + 1));
+	zassert_equal(2, dns_cache_find(&test_dns_cache, query, info_read, 3));
+	zassert_equal(AF_INET, info_read[0].ai_family);
+	k_sleep(K_MSEC(TEST_DNS_CACHE_DEFAULT_TTL * 1000 + 1));
+	zassert_equal(1, dns_cache_find(&test_dns_cache, query, info_read, 3));
+	zassert_equal(AF_INET, info_read[0].ai_family);
+	k_sleep(K_MSEC(1));
```
@jukkar 

So that some time can pass between the two calls to `dns_cache_find` and the found entry count should still remain the same since not enough time has passed.

I can remove it if you want in this particular case i don't think it makes much of a difference its only semantically more correct in my opinion.

jukkar replied:
These sleeps seem to wait for expiration, so that looks like something we cannot avoid. Could we make the sleeps smaller, like <1sec?

kica-z replied:
The problem is, that the TTL in DNS is expressed in seconds. We could internally store it in milliseconds and then adjust the test accordingly. Do you think this would be a good idea? It is an antipattern to adjust code to better fit the tests but there would be a significant test time decrease.

Alternatively is there a way to speed up time somehow?

rlubos replied:
We could limit the testing scope to `native_sim` platform only, which does not execute tests in real time (with `integration_platforms` option). Thoughts?

kica-z replied:
Sounds good to me 👍 

@jukkar @rlubos  I have adjusted the `testcase.yaml` accordingly and the checks are green.
